File: core/tarih.py
"""Türkçe tarih yardımcıları.

Vardiya listesinin sayfa adı, tarihin ayın kaçıncı haftasına düştüğünden
türetilir: 17 Ağustos -> "Ağustos3.Hafta".
"""

from __future__ import annotations

import re
from datetime import date

AYLAR = (
    "Ocak", "Şubat", "Mart", "Nisan", "Mayıs", "Haziran",
    "Temmuz", "Ağustos", "Eylül", "Ekim", "Kasım", "Aralık",
)


def coz(deger: str | date | None) -> date | None:
    """'GG.AA.YYYY' veya 'YYYY-AA-GG' biçimindeki metni tarihe çevirir."""
    if isinstance(deger, date):
        return deger
    if not deger:
        return None

    metin = str(deger).strip()
    m = re.fullmatch(r"(\d{1,2})[.\-/](\d{1,2})[.\-/](\d{4})", metin)
    if m:
        gun, ay, yil = (int(x) for x in m.groups())
    else:
        m = re.fullmatch(r"(\d{4})[.\-/](\d{1,2})[.\-/](\d{1,2})", metin)
        if not m:
            return None
        yil, ay, gun = (int(x) for x in m.groups())

    try:
        return date(yil, ay, gun)
    except ValueError:
        return None


def ayin_haftasi(tarih: date) -> int:
    """Tarihin ayın kaçıncı haftasına düştüğünü döndürür (1-5).

    Ayın ilk 7 günü 1. hafta, sonraki 7 günü 2. hafta biçiminde sayılır.
    17 Ağustos -> 3. hafta.
    """
    return (tarih.day - 1) // 7 + 1


def hafta_adi(deger: str | date | None) -> str:
    """Excel sayfa adı için 'Ağustos3.Hafta' biçiminde metin üretir.

    Tarih çözülemezse bugünün tarihi kullanılır.
    """
    tarih = coz(deger) or date.today()
    return f"{AYLAR[tarih.month - 1]}{ayin_haftasi(tarih)}.Hafta"

File: web/lib/tarih.ts
/** ISO (YYYY-AA-GG) -> Türkçe (GG.AA.YYYY). Boş girdi boş döner. */
export function ggaayyyy(iso: string): string {
  if (!iso) return "";
  const [y, a, g] = iso.split("-");
  return y && a && g ? `${g}.${a}.${y}` : "";
}

/** Türkçe (GG.AA.YYYY) -> ISO (YYYY-AA-GG). */
export function isoya(tr: string): string {
  if (!tr?.includes(".")) return "";
  const [g, a, y] = tr.split(".");
  return y && a && g ? `${y}-${a}-${g}` : "";
}

export function bugun(): string {
  const d = new Date();
  const p = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())}`;
}

const AYLAR = [
  "Ocak", "Şubat", "Mart", "Nisan", "Mayıs", "Haziran",
  "Temmuz", "Ağustos", "Eylül", "Ekim", "Kasım", "Aralık",
];

/** core/tarih.py ile aynı kural: 17 Ağustos -> "Ağustos3.Hafta". */
export function haftaAdi(iso: string): string {
  const d = iso ? new Date(iso) : new Date();
  if (Number.isNaN(d.getTime())) return "";
  return `${AYLAR[d.getMonth()]}${Math.floor((d.getDate() - 1) / 7) + 1}.Hafta`;
}
